controller/sharding/cache.go

```diff
@@ -161,3 +173,60 @@ func (d *ClusterSharding) GetClusterAccessor() clusterAccessor {
 		return clusters
 	}
 }
+
+func (d *ClusterSharding) GetAppAccessor() appAccessor {
+	return func() []*v1alpha1.Application {
+		apps := make([]*v1alpha1.Application, 0, len(d.Apps))
```

Comment:
I believe this method should acquire a read lock from ClusterSharding.

Reply:
I noticed that GetClusterAccessor doesnt have the read lock and tracing through the implementation, I'm wondering if adding a read lock could cause a deadlock. As updateDistribution calls getClusterShard, the read lock is acquired from the Add/Update/Delete mutations for clusters/app.

Reply:
OK, I haven't traced it down as far. If that's the case, I think we should make this function private and document (as in, comment) that the lock needs to be acquired elsewhere.

Reply:
Looks like cluster.go from the admin commands utilizes GetClusterAccessor/GetAppAccessor to get the destribution function. Though I think those lines of code can be removed as it doesnt do anything with the output of the destribution function. 

@akram wondering if you can take a look at this call in commands/admin/cluster.go:
```
distributionFunction := sharding.GetDistributionFunction(clusterSharding.GetClusterAccessor(), clusterSharding.GetAppAccessor(), common.DefaultShardingAlgorithm, replicas)
distributionFunction(&cluster)
```
If we can remove these lines we can convert both the GetClusterAccessor and GetAppAccessor to private.

Reply:
I did comment those lines out and tested the cli, without any issue. So if @akram does not have any objection, I will leave them commented out and change both the GetClusterAccessor  and GetAppAccessor to private.

Reply:
thanks for this great contribution @Enclavet .
supposedly,  the calls in `commands/admin/cluster.go`  are there to display the distribution using the CLI and to get the same results to what is running in the cluster.
So, if you mean that the call to these functions does not have any influence of this output, so I am more willing to think that there is a issue. Or I am getting things wrong?

Reply:
Yeah actually those calls were redundant. The admin command gets shard distribution data further up in these lines:
```
clusterShardingCache := sharding.NewClusterSharding(argoDB, shard, replicas, shardingAlgorithm)
clusterShardingCache.Init(clustersList, appItems)
clusterShards := clusterShardingCache.GetDistribution()
```

I wonder if those calls to the distributionFunction were old code that wasnt cleaned up as calling the distributionFunction directly would return the shard number of the cluster. But the return value is not stored at all.

Reply:
Converted both GetClusterAccessor and GetAppAccessor to private, added comments that both should get a read lock before calling.

Reply:
> 
> I wonder if those calls to the distributionFunction were old code that wasnt cleaned up as calling the distributionFunction directly would return the shard number of the cluster. But the return value is not stored at all.



> Yeah actually those calls were redundant. The admin command gets shard distribution data further up in these lines:
> 
> ```
> clusterShardingCache := sharding.NewClusterSharding(argoDB, shard, replicas, shardingAlgorithm)
> clusterShardingCache.Init(clustersList, appItems)
> clusterShards := clusterShardingCache.GetDistribution()
> ```
> 
> I wonder if those calls to the distributionFunction were old code that wasnt cleaned up as calling the distributionFunction directly would return the shard number of the cluster. But the return value is not stored at all.

I believe that it is old code not cleaned up. So, that's ok to remove the redundant calls.